src/Common/src/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -484,6 +486,119 @@ private static SizeAndAlignment ComputeInstanceSize(MetadataType type, int count
+                    // These are the primitive types that constitute a HFA type.
+                    return ValueTypePassingCharacteristics.HomogenousFloatAggregate;
+
+                case TypeFlags.ValueType:
+                    DefType expectedElementType = null;
+
+                    foreach (FieldDesc field in metadataType.GetFields())
+                    {
+                        if (field.IsStatic)
+                            continue;
+
+                        // If a field isn't a DefType, then this type cannot be an HFA type
+                        // If a field isn't a HFA type, then this type cannot be an HFA type
+                        DefType fieldType = field.FieldType as DefType;
+                        if (fieldType == null || !fieldType.IsHfa)
+                            return ValueTypePassingCharacteristics.None;
+
+                        if (expectedElementType == null)
+                        {
+                            // If we hadn't yet figured out what form of HFA this type might be, we've
+                            // now found one case.
+                            expectedElementType = fieldType.HfaElementType;
+                            Debug.Assert(expectedElementType != null);
+                        }
+                        else if (expectedElementType != fieldType.HfaElementType)
+                        {
+                            // If we had already determined the possible HFA type of the current type, but
+                            // the field we've encountered is not of that type, then the current type cannot
+                            // be an HFA type.
+                            return ValueTypePassingCharacteristics.None;
+                        }
+                    }
+
+                    // No fields means this is not HFA.
+                    if (expectedElementType == null)
+                        return ValueTypePassingCharacteristics.None;
+
+                    // Note that we check the total size, but do not perform any checks on number of fields:
+                    // - Type of fields can be HFA valuetype itself
+                    // - Managed C++ HFA valuetypes have just one <alignment member> of type float to signal that 
+                    //   the valuetype is HFA and explicitly specified size
+                    int maxSize = expectedElementType.InstanceFieldSize * expectedElementType.Context.Target.MaximumHfaElementCount;
+                    if (type.InstanceFieldSize > maxSize)
+                        return ValueTypePassingCharacteristics.None;
+
+                    // All the tests passed. This is an HFA type.
+                    return ValueTypePassingCharacteristics.HomogenousFloatAggregate;
+            }
+
+            return ValueTypePassingCharacteristics.None;
+        }
+
+        public override DefType ComputeHomogeneousFloatAggregateElementType(DefType type)
+        {
+            if (!type.IsHfa)
```

Comment:
In the one place where this is called, we assert that this does not happen. Remove the assert at the callsite?